File: scripts/okteto_usage.py
import re
from typing import Tuple, Iterable
import sys
from kubernetes.config import load_kube_config
from kubernetes.client import CustomObjectsApi


class CPUQuantity(object):

    def __init__(self, quantity: int = 0, unit: str = "n"):
        self.quantity = quantity
        self.unit = unit

    def convert_to_cpus(self):
        conversor_rate = {"n": 1000000000, "m": 1000}
        self.quantity = self.quantity / conversor_rate.get(self.unit)
        self.unit = "CPU"
        return self

    def __add__(self, other):
        if self.unit == other.unit:
            return CPUQuantity(self.quantity + other.quantity)
        else:
            raise ValueError()

    def __iadd__(self, other):
        if self.unit == other.unit:
            self.quantity += other.quantity
            return self
        else:
            raise ValueError()

    def __str__(self):
        return f"{self.quantity}{self.unit}"


class MemoryQuantity(object):

    def __init__(self, quantity: int = 0, unit: str = "Ki"):
        self.quantity = quantity
        self.unit = unit

    def convert_to_Mb(self):
        conversor_rate = {"Ki": 1024}
        self.quantity = self.quantity / conversor_rate.get(self.unit)
        self.unit = "MiB"
        return self

    def __add__(self, other):
        if self.unit == other.unit:
            return MemoryQuantity(self.quantity + other.quantity, self.unit)
        else:
            raise ValueError()

    def __iadd__(self, other):
        if self.unit == other.unit:
            self.quantity += other.quantity
            return self
        else:
            raise ValueError()

    def __str__(self):
        return f"{self.quantity}{self.unit}"


class PodMetrics(object):

    def __init__(self, name: str, memory_usage: MemoryQuantity, cpu_usage: CPUQuantity, number_of_containers: int):
        self.name = name
        self.memory_usage = memory_usage.convert_to_Mb()
        self.cpu_usage = cpu_usage.convert_to_cpus()
        self.number_of_containers = number_of_containers

    def __str__(self):
        return f"{self.name} => Number of containers in pod: {self.number_of_containers}" \
               f"\n{' ' * len(self.name)} ╚> Memory: {str(self.memory_usage)}\n" \
               f"{' ' * len(self.name)} ╚> CPU: {str(self.cpu_usage)}\n"


def __split_quantity_and_unit_from_string(string: str) -> Tuple[int, str]:
    temp = re.compile("([0-9]+)([a-zA-Z]+)")
    quantity, unit = temp.match(string).groups()
    return int(quantity), unit


def __get_cpu_usage_from_string(usage_string: str) -> CPUQuantity:
    quantity, unit = __split_quantity_and_unit_from_string(usage_string)
    return CPUQuantity(quantity, unit)


def __get_memory_usage_from_string(usage_string: str) -> MemoryQuantity:
    quantity, unit = __split_quantity_and_unit_from_string(usage_string)
    return MemoryQuantity(quantity, unit)


def __get_pods_metrics_from_response(results: dict) -> Iterable[PodMetrics]:
    for pod in results["items"]:
        pod_name = pod["metadata"]["name"]
        total_cpu = CPUQuantity()
        total_memory = MemoryQuantity()
        containers = pod["containers"]
        for container in containers:
            total_cpu += __get_cpu_usage_from_string(container["usage"]["cpu"])
            total_memory += __get_memory_usage_from_string(container["usage"]["memory"])
        yield PodMetrics(pod_name, total_memory, total_cpu, len(containers))


if __name__ == '__main__':
    load_kube_config()
    custom_api = CustomObjectsApi()
    results = custom_api.list_namespaced_custom_object('metrics.k8s.io', 'v1beta1', sys.argv[1], 'pods')
    for pod in __get_pods_metrics_from_response(results):
        print(pod)

File: scripts/okteto_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"

	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
	metricstype "k8s.io/metrics/pkg/apis/metrics/v1beta1"
	"k8s.io/metrics/pkg/client/clientset/versioned"
	metricsclientset "k8s.io/metrics/pkg/client/clientset/versioned"
)

type Pod struct {
	name               string
	cpuUsage           string
	memoryUsage        string
	numberOfContainers int
}

func (p Pod) String() string {
	nameLength := strings.Repeat(" ", len(p.name))
	return fmt.Sprintf("%s => Number of containers in pod: %d\n%s ╚> Memory: %s\n%s ╚> CPU: %s",
		p.name, p.numberOfContainers, nameLength, p.memoryUsage, nameLength, p.cpuUsage)
}

func main() {
	// Definition of variables to use later
	var config *rest.Config
	var namespace string
	var pods metricstype.PodMetricsList

	config = getConfig()
	namespace = "jlopezbarb"

	pods = getRawMetrics(config, namespace)

	for _, pod := range pods.Items {
		name := pod.ObjectMeta.Name
		cpuUsage, memoryUsage := getUsages(pod.Containers)

		var p Pod
		p.name = name
		p.cpuUsage = cpuUsage.String()
		p.memoryUsage = memoryUsage.String()
		p.numberOfContainers = len(pod.Containers)
		fmt.Println(p.String())
	}
}

func getUsages(containers []metricstype.ContainerMetrics) (resource.Quantity, resource.Quantity) {
	var totalCpuUsage resource.Quantity
	var totalMemoryUsage resource.Quantity
	for idx, container := range containers {
		var cpu resource.Quantity
		cpu = *container.Usage.Cpu()

		var memory resource.Quantity
		memory = *container.Usage.Memory()
		if idx == 0 {
			totalCpuUsage = cpu
			totalMemoryUsage = memory
		} else {
			totalCpuUsage.Add(cpu)
			totalMemoryUsage.Add(memory)
		}
	}
	return totalCpuUsage, totalMemoryUsage
}

func getRawMetrics(config *rest.Config, namespace string) metricstype.PodMetricsList {
	clientset, err := metricsclientset.NewForConfig(config)
	if err != nil {
		log.Fatal(err)
	}
	podMetricsList, err := clientset.MetricsV1beta1().PodMetricses(namespace).List(context.TODO(), metav1.ListOptions{})
	if err != nil {
		log.Fatal(err)
	}

	pods := *podMetricsList
	return pods
}

func getConfig() *rest.Config {
	kubeconfig := filepath.Join(
		os.Getenv("HOME"), ".kube", "config",
	)
	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
	if err != nil {
		log.Fatal(err)
	}
	return config
}

func getNumberOfRunningPods(namespace string, config *rest.Config) int {
	clientset, _ := kubernetes.NewForConfig(config)
	pods, _ := clientset.CoreV1().Pods(namespace).List(context.TODO(), metav1.ListOptions{})
	return len(pods.Items)
}

func getMetricsClient(config *rest.Config) *versioned.Clientset {
	metricsClient, err := metricsclientset.NewForConfig(config)
	if err != nil {
		log.Fatal(err)
	}
	return metricsClient
}
